modules/caddyhttp/reverseproxy/caddyfile.go

```diff
@@ -892,8 +997,76 @@ func (h *HTTPTransport) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
 	return nil
 }
 
+// Parse the tokens of a named response matcher.
+//
+//     @name {
+//         header <field> [<value>]
+//         status <code...>
+//     }
+//
+// Or, single line syntax:
+//
+//     @name [header <field> [<value>]] | [status <code...>]
+//
+func (h *Handler) parseNamedResponseMatcher(d *caddyfile.Dispenser, matchers map[string]caddyhttp.ResponseMatcher) error {
+	for d.Next() {
+		definitionName := d.Val()
+
+		if _, ok := matchers[definitionName]; ok {
+			return d.Errf("matcher is defined more than once: %s", definitionName)
+		}
+
+		matcher := caddyhttp.ResponseMatcher{}
+		for nesting := d.Nesting(); d.NextArg() || d.NextBlock(nesting); {
+			switch d.Val() {
+			case "header":
+				if matcher.Headers == nil {
+					matcher.Headers = http.Header{}
+				}
+
+				// reuse the header request matcher's unmarshaler
+				headerMatcher := caddyhttp.MatchHeader(matcher.Headers)
+				err := headerMatcher.UnmarshalCaddyfile(d.NewFromNextSegment())
+				if err != nil {
+					return err
+				}
+
+				matcher.Headers = http.Header(headerMatcher)
+
+			case "status":
+				if matcher.StatusCode == nil {
+					matcher.StatusCode = []int{}
+				}
+
+				args := d.RemainingArgs()
+				if len(args) == 0 {
+					return d.ArgErr()
+				}
+
+				for _, arg := range args {
+					if len(arg) == 3 && strings.HasSuffix(arg, "xx") {
+						arg = arg[:1]
+					}
+					statusNum, err := strconv.Atoi(arg)
+					if err != nil {
+						return d.Errf("bad status value '%s': %v", arg, err)
+					}
+					matcher.StatusCode = append(matcher.StatusCode, statusNum)
+				}
+
+			default:
+				return d.Errf("unrecognized response matcher %s", d.Val())
+			}
+		}
+
+		matchers[definitionName] = matcher
+	}
+	return nil
+}
+
+const matcherPrefix = "@"
+
 // Interface guards
 var (
-	_ caddyfile.Unmarshaler = (*Handler)(nil)
```

Comment:
This does make me a little sad. I don't know if there's any way we can keep this...

Reply:
Yeah, same. I don't really see a way. We need `Helper` here, and not `Dispenser`, and `Helper` is a superset of `Dispenser`, but the unmarshaler uses `Dispenser` :cry:

Reply:
So, I think there is a way to do this, by keeping `UnmarshalCaddyfile()` almost exactly the same, except that tokens for `handle_response` are collected and stored in the `Handler`. Then the `parseCaddyfile()` function, which calls `UnmarshalCaddyfile()`, could take those and finish parsing them out using its `Helper`.

Note that if we do remove this interface support, then any call to `UnmarshalModule` on a reverse proxy handler will break. I don't know that is currently done (it's not in our code base), but this is a breaking change otherwise.

Reply:
Hmm, I'll give it a shot. That would mean that any call to `UnmarshalCaddyfile()` that doesn't come from something which has `Helper` will just have the `handle_response` stuff dropped silently though, but that's probably better than not having `reverse_proxy` work at all. I'll play a bit with it this evening.

Reply:
Got it!!